bob-backend/src/pearl/group.rs

```diff
@@ -307,17 +322,116 @@ impl Group {
+        let mut result = false;
         for (ind, &key) in keys.iter().enumerate() {
             for (_, Leaf { data: holder, .. }) in holders.iter_possible_childs_rev(&Key::from(key))
             {
-                if !exist[ind] {
-                    exist[ind] = holder.exist(key).await.unwrap_or(false);
+                match holder.exist(key).await.unwrap_or(ReadResult::NotFound) {
+                    ReadResult::Found(ts) => {
+                        let ts = ts.into();
+                        if ts > max_timestamp.unwrap_or(0) {
+                            max_timestamp = Some(ts);
+                            result = true;
+                        }
+                    }
+                    ReadResult::Deleted(ts) => {
+                        let ts = ts.into();
+                        if ts > max_timestamp.unwrap_or(0) {
+                            max_timestamp = Some(ts);
+                            result = false;
+                        }
+                    }
+                    ReadResult::NotFound => continue,
                 }
             }
+            exist[ind] = result;
         }
         exist
     }
 
+
+    pub async fn delete(
+        &self,
+        key: BobKey,
+        meta: &BobMeta,
+        timestamp_config: StartTimestampConfig,
+        force_delete: bool,
+    ) -> Result<u64, Error> {
+        let mut reference_timestamp = meta.timestamp();
+        let mut total_deletion_count = 0;
+
+        if force_delete {
+            let actual_holder = self.get_actual_holder(get_current_timestamp(), timestamp_config).await?;
+            reference_timestamp = actual_holder.1.start_timestamp();
+            total_deletion_count += self.delete_in_actual_holder(actual_holder, key, meta).await?;
+        }
+
+        let all_holders_before: Vec<_> = {
+            let holders = self.holders.read().await;
+            holders.iter().cloned().enumerate().filter(|h| h.1.start_timestamp() < reference_timestamp).collect()
+        };
+
+        total_deletion_count += self.delete_in_holders_before(all_holders_before, key, meta).await.unwrap_or(0);
+
+        return Ok(total_deletion_count);
+    }
+
+    async fn delete_in_actual_holder(&self, holder: (ChildId, Holder), key: BobKey, meta: &BobMeta) -> Result<u64, Error> {
+        // In actual holder we delete with force_delete = true
+        let delete_count = Self::delete_common(holder.1.clone(), key, meta, true).await?;
+            // We need to add marker record to alien regardless of record presence
+        self.holders
+            .write()
+            .await
+            .add_to_parents(holder.0, &Key::from(key));
+        
+        Ok(delete_count)
+    }
+
+    async fn delete_in_holders_before(
+        &self,
+        holders: Vec<(ChildId, Holder)>,
+        key: BobKey,
+        meta: &BobMeta
+    ) -> Result<u64, Error> {
+        let mut total_count = 0;
+        // Reverse iteration (from newer to older)
```

Comment:
Why do we care about order of iteration here?

Reply:
Actually, the order is not important here. It's just a general assumption that it is better to remove in order from new holders to old ones. I'll expand the comment to indicate that this behavior is not strictly required